<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Humburg" />
  <title>Introduction to eQTL analysis</title>
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no">
  <link rel="stylesheet" href="reveal.js/css/reveal.min.css"/>
    <style type="text/css">code{white-space: pre;}</style>
    <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; }
code > span.dt { color: #204a87; }
code > span.dv { color: #0000cf; }
code > span.bn { color: #0000cf; }
code > span.fl { color: #0000cf; }
code > span.ch { color: #4e9a06; }
code > span.st { color: #4e9a06; }
code > span.co { color: #8f5902; font-style: italic; }
code > span.ot { color: #8f5902; }
code > span.al { color: #ef2929; }
code > span.fu { color: #000000; }
code > span.er { font-weight: bold; }
    </style>
    <link rel="stylesheet" href="include/slides.css"/>
    <!-- If the query includes 'print-pdf', include the PDF print sheet -->
    <script>
      if( window.location.search.match( /print-pdf/gi ) ) {
        var link = document.createElement( 'link' );
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = 'reveal.js/css/print/pdf.css';
        document.getElementsByTagName( 'head' )[0].appendChild( link );
      }
    </script>
    <!--[if lt IE 9]>
    <script src="reveal.js/lib/js/html5shiv.js"></script>
    <![endif]-->
    <script src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section>
    <h1 class="title">Introduction to eQTL analysis</h1>
    <h2 class="author">Peter Humburg</h2>
    <h3 class="date"></h3>
</section>
<section id="TOC">
<h1>Overview</h1>
<ul>
<li><a href="#/what-are-eqtl">What are eQTL?</a></li>
<li><a href="#/detecting-eqtl-a-simple-model">Detecting eQTL – A simple model</a></li>
<li><a href="#/hands-on-simple-linear-regression">Hands-on: Simple linear regression</a></li>
</ul>
</section>

<section><section id="what-are-eqtl" class="titleslide slide level1"><h1>What are eQTL?</h1></section></section>
<section><section id="detecting-eqtl-a-simple-model" class="titleslide slide level1"><h1>Detecting eQTL – A simple model</h1></section><section id="how-do-we-know-whether-a-locus-is-associated-with-the-expression-of-a-gene" class="slide level2" data-transition="none">
<h1>How do we know whether a locus is associated with the expression of a gene?</h1>
<div class="multicolumn">
<div>
<figure>
<img src="figure/genomic_locus.png" />
</figure>
</div>
<div>
<ul>
<li><strong>Any given locus contains multiple SNPs.</strong></li>
<li>Can determine genotypes for these SNPs in a (large) number of individuals.</li>
<li>Measure gene expression for genes of interest.</li>
<li>Assess the evidence that expression varies with genotype.</li>
</ul>
</div>
</div>
</section><section id="how-do-we-know-whether-a-locus-is-associated-with-the-expression-of-a-gene-1" class="slide level2" data-transition="none">
<h1>How do we know whether a locus is associated with the expression of a gene?</h1>
<div class="multicolumn">
<div>
<table>
<thead>
<tr class="header">
<th style="text-align: center;">sample</th>
<th style="text-align: left;">snp_1</th>
<th style="text-align: left;">snp_2</th>
<th style="text-align: left;">snp_3</th>
<th style="text-align: center;">…</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">sample 1</td>
<td style="text-align: left;">AA</td>
<td style="text-align: left;">AA</td>
<td style="text-align: left;">AB</td>
<td style="text-align: center;">…</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample 2</td>
<td style="text-align: left;">AB</td>
<td style="text-align: left;">AB</td>
<td style="text-align: left;">AA</td>
<td style="text-align: center;">…</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample 3</td>
<td style="text-align: left;">AB</td>
<td style="text-align: left;">BB</td>
<td style="text-align: left;">AB</td>
<td style="text-align: center;">…</td>
</tr>
<tr class="even">
<td style="text-align: center;">…</td>
<td style="text-align: left;">…</td>
<td style="text-align: left;">…</td>
<td style="text-align: left;">…</td>
<td style="text-align: center;">…</td>
</tr>
</tbody>
</table>
</div>
<div>
<ul>
<li>Any given locus contains multiple SNPs.</li>
<li><strong>Can determine genotypes for these SNPs in a (large) number of individuals.</strong></li>
<li>Measure gene expression for genes of interest.</li>
<li>Assess the evidence that expression varies with genotype.</li>
</ul>
</div>
</div>
</section><section id="how-do-we-know-whether-a-locus-is-associated-with-the-expression-of-a-gene-2" class="slide level2" data-transition="none">
<h1>How do we know whether a locus is associated with the expression of a gene?</h1>
<div class="multicolumn">
<div>
<table>
<thead>
<tr class="header">
<th style="text-align: center;">sample</th>
<th style="text-align: left;">gene_1</th>
<th style="text-align: right;">gene_2</th>
<th style="text-align: left;">gene_3</th>
<th style="text-align: center;">…</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: center;">sample 1</td>
<td style="text-align: left;">7.3</td>
<td style="text-align: right;">12.8</td>
<td style="text-align: left;">6.5</td>
<td style="text-align: center;">…</td>
</tr>
<tr class="even">
<td style="text-align: center;">sample 2</td>
<td style="text-align: left;">10.9</td>
<td style="text-align: right;">9.6</td>
<td style="text-align: left;">8.8</td>
<td style="text-align: center;">…</td>
</tr>
<tr class="odd">
<td style="text-align: center;">sample 3</td>
<td style="text-align: left;">9.5</td>
<td style="text-align: right;">10.7</td>
<td style="text-align: left;">15.1</td>
<td style="text-align: center;">…</td>
</tr>
<tr class="even">
<td style="text-align: center;">…</td>
<td style="text-align: left;">…</td>
<td style="text-align: right;">…</td>
<td style="text-align: left;">…</td>
<td style="text-align: center;">…</td>
</tr>
</tbody>
</table>
</div>
<div>
<ul>
<li>Any given locus contains multiple SNPs.</li>
<li>Can determine genotypes for these SNPs in a (large) number of individuals.</li>
<li><strong>Measure gene expression for genes of interest.</strong></li>
<li>Assess the evidence that expression varies with genotype.</li>
</ul>
</div>
</div>
</section><section id="how-do-we-know-whether-a-locus-is-associated-with-the-expression-of-a-gene-3" class="slide level2" data-transition="none">
<h1>How do we know whether a locus is associated with the expression of a gene?</h1>
<div class="multicolumn">
<div id="snp-expr">
<img src="figure/snp_expression.png" />
</div>
<div>
<ul>
<li>Any given locus contains multiple SNPs.</li>
<li>Can determine genotypes for these SNPs in a (large) number of individuals.</li>
<li>Measure gene expression for genes of interest.</li>
<li><strong>Assess the evidence that expression varies with genotype.</strong></li>
</ul>
</div>
</div>
</section><section id="linear-additive-model" class="slide level2">
<h1>Linear additive model</h1>
<div class="multicolumn">
<div id="snp-expr">
<img src="figure/snp_expression.png" />
</div>
<div>
<p>Different alleles of a SNP may exhibit a dosage effect.</p>
<ul>
<li>Using the <em>AA</em> genotype as baseline, each copy of the <em>B</em> allele changes expression by a fixed amount.</li>
<li>Implies a linear relationship between the mean gene expression and the number of <em>B</em> alleles.</li>
<li>Estimating the change in expression due to the <em>B</em> allele to quantify the SNP’s contribution to gene expression.</li>
</ul>
</div>
</div>
</section><section id="linear-regression" class="slide level2" data-transition="none">
<h1>Linear regression</h1>
<p><span class="math">\[Y = \beta_0 + \beta X + \varepsilon\]</span></p>
<dl>
<dt>Y</dt>
<dd>Response variable (here: vector of expression values for gene of interest).
</dd>
<dt>X</dt>
<dd>Explanatory variable (here: vector of genotypes (coded as 0, 1, 2) for the SNP under consideration).
</dd>
</dl>
</section><section id="linear-regression-1" class="slide level2" data-transition="none">
<h1>Linear regression</h1>
<p><span class="math">\[Y = \beta_0 + \beta X + \varepsilon\]</span></p>
<dl>
<dt><span class="math">\(\beta_0\)</span></dt>
<dd>Intercept (here: mean expression for <em>AA</em> genotype).
</dd>
<dt><span class="math">\(\beta\)</span></dt>
<dd>Regression coefficient; the effect of X on the mean of Y (here: change in mean gene expression for each copy of the <em>B</em> allele).
</dd>
<dt><span class="math">\(\varepsilon\)</span></dt>
<dd>Residuals; the difference between observed values of <span class="math">\(Y\)</span> and the estimated mean of <span class="math">\(Y|X\)</span>
</dd>
</dl>
</section><section id="linear-regression-assumptions" class="slide level2">
<h1>Linear regression – Assumptions</h1>
<p>Residuals are</p>
<ul>
<li>independent</li>
<li>normally distributed with mean 0 and constant variance</li>
</ul>
<div class="fragment">
<p>It is implied that</p>
<ul>
<li>Values of <span class="math">\(Y\)</span> for each value of <span class="math">\(X\)</span> are normally distributed.</li>
<li>There is a linear relationship between <span class="math">\(X\)</span> and <span class="math">\(Y\)</span>.</li>
</ul>
</div>
</section><section id="linear-regression-robustness" class="slide level2" data-transition="none">
<h1>Linear regression – Robustness</h1>
<h3 id="independence-of-residuals">Independence of residuals</h3>
<p>Can produce misleading results.</p>
<p>What could cause this?</p>
</section><section id="linear-regression-robustness-1" class="slide level2" data-transition="none">
<h1>Linear regression – Robustness</h1>
<h3 id="constant-variance-of-residuals-homoskedacity">Constant variance of residuals (homoskedacity)</h3>
<p>Violation of this assumption will lead to incorrect p-values and confidence intervals.</p>
</section><section id="linear-regression-robustness-2" class="slide level2" data-transition="none">
<h1>Linear regression – Robustness</h1>
<h3 id="normality">Normality</h3>
<ul>
<li>Estimates and their confidence intervals and p-values are fairly robust.</li>
<li>Prediction can become problematic (but we are not interested in that here).</li>
</ul>
</section><section id="linear-regression-robustness-3" class="slide level2" data-transition="none">
<h1>Linear regression – Robustness</h1>
<h3 id="linearity">Linearity</h3>
<div class="left">
<p>If the true relationship between <span class="math">\(Y\)</span> and <span class="math">\(X\)</span> is non-linear conclusions may be misleading.</p>
When might this occur with eQTL data?
</div>
</section><section id="linear-regression-robustness-4" class="slide level2" data-transition="none">
<h1>Linear regression – Robustness</h1>
<h3 id="linearity-1">Linearity</h3>
<div class="left">
<p>If the true relationship between <span class="math">\(Y\)</span> and <span class="math">\(X\)</span> is non-linear conclusions may be misleading.</p>
When might this occur with eQTL data?
</div>
<div class="right">
<img src="figure/dominant.png" />
</div>
</section></section>
<section><section id="hands-on-simple-linear-regression" class="titleslide slide level1"><h1>Hands-on: Simple linear regression</h1></section><section id="setup" class="slide level2">
<h1>Setup</h1>
<ul>
<li><p>Make sure you have the latest version of the Docker image</p>
<pre class="sh"><code>docker pull humburg/eqtl-intro</code></pre></li>
<li><p>Start the RStudio server</p>
<pre class="sh"><code>docker run -p 8787:8787 humburg/eqtl-intro</code></pre></li>
<li>On <strong>Mac</strong> and <strong>Windows</strong> determine the servers IP address with <code>boot2docker ip</code>. On <strong>Linux</strong> use <em>localhost</em>.</li>
<li>Access the RStudio interface at <em>http://yourip:8787</em>.
<ul>
<li>Username: rstudio</li>
<li>Password: rstudio</li>
</ul></li>
</ul>
</section><section id="data" class="slide level2">
<h1>Data</h1>
<dl>
<dt>Genotypes</dt>
<dd><em>/data/simulated/sim_genotypes.tab</em>
</dd>
<dt>Gene expression</dt>
<dd><em>/data/simulated/expression1.tab</em>
</dd>
</dl>
</section><section id="warm-up" class="slide level2">
<h1>Warm-up</h1>
<ol>
<li class="fragment">Load the data into R.</li>
<li class="fragment">How is the data formatted?</li>
<li class="fragment">Determine minor allele frequencies.</li>
</ol>
<aside class="notes">
<pre class="sourceCode r"><code class="sourceCode r">geno &lt;-<span class="st"> </span>readr::<span class="kw">read_tsv</span>(<span class="st">&quot;/data/simulated/sim_genotypes.tab&quot;</span>)
expr &lt;-<span class="st"> </span>readr::<span class="kw">read_tsv</span>(<span class="st">&quot;/data/simulated/sim_expression1.tab&quot;</span>)

maf &lt;-<span class="st"> </span><span class="kw">colMeans</span>(geno[-<span class="dv">1</span>])/<span class="dv">2</span></code></pre>
</aside>
</section><section id="plotting-the-data" class="slide level2">
<h1>Plotting the data</h1>
<ul>
<li>Choose a SNP/gene pair (snp_1 / gene_1, snp_2 / gene_2, …)</li>
<li>Create a plot showing gene expression by genotype for this pair.</li>
</ul>
<aside class="notes">
<pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(ggplot2)
genoLong &lt;-<span class="st"> </span>tidyr::<span class="kw">gather</span>(geno, snp, genotype, -sample)
exprLong &lt;-<span class="st"> </span>tidyr::<span class="kw">gather</span>(expr, gene, expression, -sample)
dataLong &lt;-<span class="st"> </span><span class="kw">cbind</span>(genoLong, exprLong[<span class="st">&quot;expression&quot;</span>])
dataLong$genotype &lt;-<span class="st"> </span><span class="kw">as.factor</span>(dataLong$genotype) 
<span class="kw">ggplot</span>(dataLong, <span class="kw">aes</span>(genotype, expression)) +
<span class="st">        </span><span class="kw">geom_jitter</span>(<span class="dt">colour=</span><span class="st">&quot;darkgrey&quot;</span>, 
                <span class="dt">position=</span><span class="kw">position_jitter</span>(<span class="dt">width=</span><span class="fl">0.25</span>)) +
<span class="st">        </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.size=</span><span class="dv">0</span>, <span class="dt">alpha=</span><span class="fl">0.6</span>, <span class="dt">fill=</span><span class="st">&quot;grey&quot;</span>) +<span class="st"> </span>
<span class="st">        </span><span class="kw">facet_wrap</span>(~snp) +<span class="st"> </span><span class="kw">theme_bw</span>()</code></pre>
</aside>
<div class="references">

</div>
</section></section>
    </div>
  </div>


  <script src="reveal.js/lib/js/head.min.js"></script>
  <script src="reveal.js/js/reveal.min.js"></script>

  <script>

      // Full list of configuration options available here:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        controls: true,
        progress: true,
        history: true,
        center: true,
        theme: Reveal.getQueryHash().theme, // available themes are in /css/theme
        transition: Reveal.getQueryHash().transition || 'default', // default/cube/page/concave/zoom/linear/fade/none

        // Optional libraries used to extend on reveal.js
        dependencies: [
          { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true, condition: function() { return !!document.body.classList; } },
          { src: 'reveal.js/plugin/notes/notes.js', async: true, condition: function() { return !!document.body.classList; } },
//          { src: 'reveal.js/plugin/search/search.js', async: true, condition: function() { return !!document.body.classList; }, }
//          { src: 'reveal.js/plugin/remotes/remotes.js', async: true, condition: function() { return !!document.body.classList; } }
]});
    </script>
    </body>
</html>
